openhands/server/conversation_manager/docker_nested_conversation_manager.py

```diff
@@ -0,0 +1,342 @@
+            conversations = await conversation_store.get_all_metadata(response_ids)
+            conversations.sort(key=_last_updated_at_key, reverse=True)
+
+            while len(conversations) >= self.config.max_concurrent_conversations:
+                oldest_conversation_id = conversations.pop().conversation_id
+                logger.debug(
+                    f'closing_from_too_many_sessions:{user_id or ""}:{oldest_conversation_id}',
+                    extra={'session_id': oldest_conversation_id, 'user_id': user_id},
+                )
+                # Send status message to client and close session.
+                status_update_dict = {
+                    'status_update': True,
+                    'type': 'error',
+                    'id': 'AGENT_ERROR$TOO_MANY_CONVERSATIONS',
+                    'message': 'Too many conversations at once. If you are still using this one, try reactivating it by prompting the agent to continue',
+                }
+                await self.sio.emit(
+                    'oh_event',
+                    status_update_dict,
+                    to=ROOM_KEY.format(sid=oldest_conversation_id),
+                )
+                await self.close_session(oldest_conversation_id)
+
+    def _get_provider_handler(self, settings: Settings):
+        provider_tokens = None
+        if isinstance(settings, ConversationInitData):
+            provider_tokens = settings.git_provider_tokens
+        provider_handler = ProviderHandler(
+            provider_tokens=provider_tokens
+            or cast(PROVIDER_TOKEN_TYPE, MappingProxyType({}))
+        )
+        return provider_handler
+
+    async def _create_runtime(self, sid: str, user_id: str, settings: Settings, provider_handler: ProviderHandler):
+
+        # This session is created here only because it is the easiest way to get a runtime, which
+        # is the easiest way to create the needed docker container
+        session = Session(
+            sid=sid,
+            file_store=self.file_store,
+            config=self.config,
+            sio=self.sio,
+            user_id=user_id,
+        )
+        agent_cls = settings.agent or self.config.default_agent
+        agent_name = agent_cls if agent_cls is not None else 'agent'
+        llm = LLM(
+            config=self.config.get_llm_config_from_agent(agent_name),
+            retry_listener=session._notify_on_llm_retry,
+        )
+        llm = session._create_llm(agent_cls)
+        agent_config = self.config.get_agent_config(agent_cls)
+        agent = Agent.get_cls(agent_cls)(llm, agent_config)
+
+        env_vars = await provider_handler.get_env_vars(expose_secrets=True)
+        env_vars['CONVERSATION_MANAGER_CLASS'] = 'openhands.server.conversation_manager.standalone_conversation_manager.StandaloneConversationManager'
+        env_vars['SERVE_FRONTEND'] = '0'
+        env_vars['RUNTIME'] = 'local'
+        env_vars['CONVERSATION_ID'] = sid
+        env_vars['USER'] = 'CURRENT_USER'
```

Comment:
This is probably fine for DockerNested, but we'll want to get these set on startup elsewhere